api/trace/span_context.go

```diff
@@ -64,6 +64,23 @@ func (t ID) MarshalJSON() ([]byte, error) {
 	return json.Marshal(t.String())
 }
 
+// UnmarshalJSON implements a custom unmarshal function to decode
+// hex string as a TraceID.
+func (t *ID) UnmarshalJSON(b []byte) error {
+	data := struct {
+		TraceID string `json:"TraceID"`
+	}{}
+	if err := json.Unmarshal(b, &data); err != nil {
+		return err
+	}
```

Comment:
This approach will only work for narrowly defined JSON (this will fail to unmarshal `{"bob": "80f198ee56343ba864fe8b2a57d3eff7"}` into ``struct {ID ID `json:"bob"`}``).

Given you already have the ID as a byte slice, it seems like a better approach to parse that slice directly. Maybe something like:

```suggestion
	// Given hex is not valid JSON, assume it is a string.
	b = bytes.Trim(b, `"`)
	if hex.DecodedLen(len(b)) != 16 {
		return ErrInvalidTraceIDLength
	}
	// Maybe don't decode directly here, but validate decode size before assigning?
	_, err := hex.Decode((*t)[:], b)
	if err != nil {
		return err
	}
	return nil
```

Reply:
^^^ I suspect that some implementations will not use a leading 0, so the len() == 16 test might not work.